src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs

```diff
@@ -573,6 +625,18 @@ private bool TryGetPropertyForInputModelProperty(Func<ObjectTypeProperty, bool>
             );
         }
 
+        private ObjectTypeProperty GetPropertyForDiscriminator(string inputPropertyName)
+        {
+            foreach (var obj in EnumerateHierarchy())
+            {
+                var property = obj.Properties.FirstOrDefault(p => p.InputModelProperty?.Name == inputPropertyName);
```

Comment:
will this ensure we pick the right one if there is a duplicate name of the properties for the discriminator?

Reply:
This is filtering through `Properties` on `ObjectType`, which we have already filtered the properties with duplicated names (see [this comment](https://github.com/Azure/autorest.csharp/pull/4097#discussion_r1449859334))